### PR TITLE
Flight properties

### PR DESCRIPTION
**Pull request by kylep-dremio**
Add optional arbitrary properties to flight calls.

### CODE REVIEW

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightConstants.java [hunk 1]
```diff
@@ -23,5 +23,6 @@
 public interface FlightConstants {
 
   String SERVICE = "arrow.flight.protocol.FlightService";
+  String PROPERTY_HEADER = "ArrowProperties";
```
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightConstants.java]: Header names are usually dash-separated.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/PropertyCallOption.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        outStream.writeObject(property.getValue());
+        value.append(encoder.encodeToString(byteStream.toByteArray()));
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      value.append(";");
+    }
+
+    propertiesMetadata = new Metadata();
+    propertiesMetadata.put(
```
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/PropertyCallOption.java]: What if the client did not send any properties?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+     * Construct a factory with the given property handler.
+     * @param propertyHandler The property handler that will be used to pass properties.
+     */
+    public Factory(ServerPropertyHandler propertyHandler) {
+      this.propertyHandler = propertyHandler;
+    }
+
+    @Override
+    public ServerPropertyMiddleware onCallStarted(CallInfo callInfo, CallHeaders incomingHeaders,
+                                                  RequestContext context) {
```
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java]: This needs to be indented to 4 spaces past the p in public. Checkstyle won't complain about this.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      final String headerVal = incomingHeaders.get(FlightConstants.PROPERTY_HEADER);
+      final Map<String, Object> properties = new HashMap<>();
+      for (String property : headerVal.split(";")) {
+        final String[] splitProp = property.split("=");
+        if (splitProp.length == 2) {
```
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java]: Throw on malformed properties?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+      if (!properties.isEmpty()) {
+        propertyHandler.accept(properties);
+      }
+
+      return new ServerPropertyMiddleware();
+    }
+  }
+
+  public ServerPropertyMiddleware() {
```
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java]: Private constructor?

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestCallOptions.java [hunk 6]
```diff
@@ -64,14 +66,29 @@ public void underTimeout() {
     });
   }
 
-  void test(Consumer<FlightClient> testFn) {
+  @Test
+  public void propertyHeader() {
```
Comment (jduo) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestCallOptions.java]: Let's add more tests with multiple properties, properties with special chars in name/value.

**Review by jduo — CHANGES_REQUESTED** [review]

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/PropertyCallOption.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Method option for supplying properties to method calls.
+ */
+public class PropertyCallOption implements CallOptions.GrpcCallOption {
+  private final Metadata propertiesMetadata;
+
+  /**
+   * Single property constructor.
+   */
+  public PropertyCallOption(String key, String value) {
```
Comment (kylep-dremio) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/PropertyCallOption.java]: Value should be a serializable.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java (hunk 5 above)
Comment (kylep-dremio) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java]: Also updated ServerCallHeaderAuthMiddleware as that's where this was taken from.

**Review by jduo — APPROVED** [review]
LGTM. Let's not merge this to flight-auth-redesign, then once that gets merged to apache/master, make the PR against that.

**Pull request closed without merging**